Sort code autocompletion with rules

ajreckof opened the pull request:
supersedes #65655

This PR sort results based on the rules following :
- if we are completing an empty string it is sort alphabetically (might be a good idea to modify that)
- it ranks higher, in this order of priority, matches : 
	1. with lesser fragment
	2. starting from the start of the string to complete
	3. closer in terms of scope
	4. with lesser mismatched case
	5. starting the closest from the start
- if it fails it will try sorting them based on all the position of matches ( going through all positions the first match matching at this position while the other doesn't will be ranked higher)
-if it still fails they are sorted alphabetically

The sort on option will first get the best match for each option based on the rules above. After that it will rank every option based on the best match found.

If you have found a case that is wrongly sorted I would gladly improve the rules. If in addition you have an idea on how to correct the rules for your use case feel free to propose a new rule.

*Bugsquad edit:*
- Fixes #63706

Mickeon commented:
Thank you for crediting and the inspiration from the very original PR's algorithm!

adamscott reviewed (commented):
The GDScript team reviewed the PR. The idea is nice, @vnen will review the code later. Maybe myself too.

Will be watching this PR closely. We'll ask the production team if it's worth it to merge it sooner rather than later in order to have feedback.

Please squash your commits in order to be able to merge it later.

Thank you!

YuriSizov commented:
> We'll ask the production team if it's worth it to merge it sooner rather than later in order to have feedback.

There is no hurry per se, but given the current state of the feature it's something we should fix for 4.1. I think we should still give it reasonable testing before merging, so we are not only dependent on arbitrary and sporadic user feedback.

We should consider adding tests for this, and then if we get more feedback for cases we didn't consider, then we add more tests. Testing this can be tricky. I think we might want to expose some more internal methods to get a list of ranked suggestions and their values for a certain context and input.

ajreckof commented:
I'll look into it. For the moment I'll populate with all examples that were given on Mickeon's PR.

vnen commented:
> We should consider adding tests for this, and then if we get more feedback for cases we didn't consider, then we add more tests. Testing this can be tricky. I think we might want to expose some more internal methods to get a list of ranked suggestions and their values for a certain context and input.

I have been thinking on how to test completion since I started the test runner, so far I could not find a solution. The main problem is that completion is a list of things in the engine, which may change at any point. I wouldn't want that a change in Node (like adding a new function) breaks the GDScript test just because the completions list changed.

So for now I wouldn't fret about testing completion for this PR in particular. If at some point we're able to figure out testing we can include this code as one thing to test. It would be nice if more people tested the PR before merging but I think it's unlikely to happen. I believe this one is already better than the current state (which isn't great) and we can tweak it later if needed.

ajreckof commented:
I do agree that it should not break just because someone added a new method somewhere. So that's why I was looking into giving him a set of options a string to complete and checking the order on this small set of options. It would only check the ordering and not really the autocompletion in itself. But as this PR is about ordering and not extracting options from environnement, it feels like that is what i should add test about here.

Paulb23 reviewed (changes requested):
We should probably update the `GDVIRTUAL_CALL(_filter_code_completion_candidates`  return and arg dictionarys to support the new property, to stop it from diverging. As it's a dictionary shouldn't break ABI compact.

ajreckof commented on doc/classes/CodeEdit.xml:
```diff
@@ -551,6 +552,12 @@
 		<constant name="KIND_PLAIN_TEXT" value="9" enum="CodeCompletionKind">
 			Marks the option as unclassified or plain text.
 		</constant>
+		<constant name="LOCATION_LOCAL" value="2" enum="CodeCompletionLocation">
+		</constant>
+		<constant name="LOCATION_BASE" value="4" enum="CodeCompletionLocation">
+		</constant>
+		<constant name="LOCATION_OTHER" value="1024" enum="CodeCompletionLocation">
+		</constant>
```
description are set in [ScriptLanguageExtension.xml](https://github.com/godotengine/godot/pull/75746/files#diff-e7754195c0c111aac322b22fe61d238c9016315c5b43d73b09bfd874b3ecc07d), there is one missing for `LOCATION_BASE` still so I'll make it.
There's something strange here those enum constants are defined at two places I'll look at why and try to resolve it

ajreckof replied:
Okay so it is wanted they are at both places. I'll just move descriptions from [ScriptLanguageExtension.xml](https://github.com/godotengine/godot/pull/75746/files#diff-e7754195c0c111aac322b22fe61d238c9016315c5b43d73b09bfd874b3ecc07d) to here.

ajreckof commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -756,7 +756,8 @@ void ScriptTextEditor::_code_complete_script(const String &p_code, List<ScriptLa
 
-	r_options->sort_custom_inplace<CodeCompletionOptionCompare>();
+	// Editor will sort later, no point doing it now?
+	//r_options->sort_custom_inplace<CodeCompletionOptionCompare>();
```
Removed

Paulb23 commented on scene/gui/code_edit.cpp:
```diff
@@ -2031,14 +2030,15 @@ void CodeEdit::request_code_completion(bool p_force) {
 	completion_option.font_color = p_text_color;
 	completion_option.icon = p_icon;
 	completion_option.default_value = p_value;
+	completion_option.location = (ScriptLanguage::CodeCompletionLocation)p_location;
```
There's also `get_code_completion_option` and `get_code_completion_options`

ajreckof replied:
thanks.

ajreckof commented on scene/gui/code_edit.cpp:
```diff
@@ -3384,3 +3323,32 @@ CodeEdit::CodeEdit() {
+bool CodeCompletionOptionCompare::operator()(const ScriptLanguage::CodeCompletionOption &l, const ScriptLanguage::CodeCompletionOption &r) const {
+	// Check if we are not completing an empty string in this case there is no reason to get matches characteristics.
+	// This enables us to assure option passed to get_option_characteristics are not null matches.
+	if (base.length() == 0) {
+		return l.display < r.display;
+	}
```
Yeah you are right I can now remove it properly. the static var was first introduced to compute characteristics but as it is now only retrieving the cached value it is not needed there anymore I should be able to ensure the comparator is never called when `base` is empty.

ajreckof replied:
so to remove it I made `get_option_characteristics` treats the case of empty base and return an empty characteristics

vnen commented on core/object/script_language.h:
```diff
@@ -321,10 +322,10 @@ class ScriptLanguage : public Object {
+		LOCATION_LOCAL = 1 << 0,
+		LOCATION_BASE = 1 << 1,
 		LOCATION_OTHER = 1 << 10,
 	};
```
Changing this enum is technically a compatibility breakage. Extensions using the current version won't work with this change.

ajreckof replied:
back tracked this change so that itwouldn't be changed on `CodeEdit` side either.

Paulb23 reviewed (approved):
Can't speak to the new sorting. 

But the `CodeEdit` and `ScriptTextEditor` changes look good to me.

akien-mga commented on core/object/script_language.cpp:
```diff
@@ -460,6 +459,51 @@ void ScriptLanguage::get_core_type_words(List<String> *p_core_type_words) const
+}
+
+TypedArray<int> ScriptLanguage::CodeCompletionOption::get_option_cached_characteristics() const {
+	// only returns the cached value and warns if it was not updated since the last change of matches
```
Please use a capital letter and a dot for comments (sentence case).

akien-mga commented on core/object/script_language.cpp:
```diff
@@ -460,6 +459,51 @@ void ScriptLanguage::get_core_type_words(List<String> *p_core_type_words) const
+TypedArray<int> ScriptLanguage::CodeCompletionOption::get_option_cached_characteristics() const {
+	// only returns the cached value and warns if it was not updated since the last change of matches
+	if (last_matches != matches) {
+		WARN_PRINT("characteristics are not up to date");
```
Likewise here, warnings should be formatted with sentence case.

akien-mga commented on core/object/script_language.cpp:
```diff
@@ -460,6 +459,51 @@ void ScriptLanguage::get_core_type_words(List<String> *p_core_type_words) const
 
 void ScriptLanguage::frame() {
 }
+TypedArray<int> ScriptLanguage::CodeCompletionOption::get_option_characteristics(String base) {
```
Keep a separation line between unrelated methods definitions.

akien-mga replied:
Parameters should use `p_` prefix, and I guess this one could be 
```suggestion
TypedArray<int> ScriptLanguage::CodeCompletionOption::get_option_characteristics(const String &p_base) {
```

akien-mga commented on core/object/script_language.cpp:
```diff
@@ -460,6 +459,51 @@ void ScriptLanguage::get_core_type_words(List<String> *p_core_type_words) const
 void ScriptLanguage::frame() {
 }
+TypedArray<int> ScriptLanguage::CodeCompletionOption::get_option_characteristics(String base) {
+	// return characacteristics of the match found by order of importance
+	// matches will be ranked by a lexicographical order on the vector returned by this function
+	// the lower values indicate better matches and that they should go before in the order of appearance
```
```suggestion
	// Return characteristics of the match found by order of importance.
	// Matches will be ranked by a lexicographical order on the vector returned by this function.
	// The lower values indicate better matches and that they should go before in the order of appearance.
```
Likewise for other comments.

akien-mga commented on core/object/script_language.h:
```diff
@@ -305,7 +306,7 @@ class ScriptLanguage : public Object {
 	virtual bool overrides_external_editor() { return false; }
 
-	/* Keep enum in Sync with:                               */
+	/* Keep enums in Sync with:                               */
```
I would suggest removing the trailing `*/` since it's now out of alignment, and doesn't add much anyway.
Probably best would be to change the comment style to `//` (and while at it change "Sync" to "sync").

akien-mga commented on doc/classes/ScriptLanguageExtension.xml:
```diff
@@ -358,16 +358,12 @@
 		<constant name="LOOKUP_RESULT_MAX" value="9" enum="LookupResultType">
 		</constant>
 		<constant name="LOCATION_LOCAL" value="0" enum="CodeCompletionLocation">
-			The option is local to the location of the code completion query - e.g. a local variable.
 		</constant>
 		<constant name="LOCATION_PARENT_MASK" value="256" enum="CodeCompletionLocation">
-			The option is from the containing class or a parent class, relative to the location of the code completion query. Perform a bitwise OR with the class depth (e.g. 0 for the local class, 1 for the parent, 2 for the grandparent, etc) to store the depth of an option in the class or a parent class.
 		</constant>
 		<constant name="LOCATION_OTHER_USER_CODE" value="512" enum="CodeCompletionLocation">
-			The option is from user code which is not local and not in a derived class (e.g. Autoload Singletons).
 		</constant>
 		<constant name="LOCATION_OTHER" value="1024" enum="CodeCompletionLocation">
-			The option is from other engine code, not covered by the other enum constants - e.g. built-in classes.
```
This were like lost by mistake?

ajreckof replied:
This was removed and put in `code_edit.xml` it felt better as this was what was done for the other shared enum

akien-mga replied:
Well this class still needs documentation too, so we shouldn't remove docs. It's not uncommon to have duplicate descriptions for duplicated APIs.

ajreckof replied:
Thought this was intended as this class does not have any documentation. Then should I also duplicate the doc for the other enum that is shared?

akien-mga commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3075,6 +3076,9 @@ ::Error GDScriptLanguage::complete_code(const String &p_code, const String &p_pa
 	}
 
 	for (const KeyValue<String, ScriptLanguage::CodeCompletionOption> &E : options) {
+		if (E.value.location != 1024) {
+			print_line(E.value.display, "\t:\t", E.value.location);
+		}
```
Debug print?

akien-mga commented on scene/gui/code_edit.h:
```diff
@@ -37,7 +37,7 @@ class CodeEdit : public TextEdit {
 
 public:
-	/* Keep enum in sync with:                                           */
+	/* Keep enums in sync with:                                           */
 	/* /core/object/script_language.h - ScriptLanguage::CodeCompletionKind */
```
```suggestion
	// Keep enums in sync with:
	// core/object/script_language.h - ScriptLanguage::CodeCompletionKind
```

akien-mga commented on scene/gui/code_edit.h:
```diff
@@ -427,7 +435,7 @@ class CodeEdit : public TextEdit {
 	void request_code_completion(bool p_force = false);
 
-	void add_code_completion_option(CodeCompletionKind p_type, const String &p_display_text, const String &p_insert_text, const Color &p_text_color = Color(1, 1, 1), const Ref<Resource> &p_icon = Ref<Resource>(), const Variant &p_value = Variant::NIL);
+	void add_code_completion_option(CodeCompletionKind p_type, const String &p_display_text, const String &p_insert_text, const Color &p_text_color = Color(1, 1, 1), const Ref<Resource> &p_icon = Ref<Resource>(), const Variant &p_value = Variant::NIL, int p_location = LOCATION_OTHER);
```
Why not use `CodeCompletionLocation` as type? Hopefully this might also properly expose the type in the docs, as currently there's no link made between that parameter and the enum.

ajreckof replied:
unfortunately this is not a `BitField` either

akien-mga commented on tests/scene/test_code_edit.h:
```diff
@@ -3445,6 +3445,98 @@ TEST_CASE("[SceneTree][CodeEdit] completion") {
 	}
 
+	SUBCASE("[CodeEdit] autocomplete suggestion order") {
+		/* favorise less fragmented suggestion */
```
```suggestion
		/* Favorize less fragmented suggestion. */
```
Sentence case and en_US spelling.

Likewise for others.

akien-mga commented on tests/scene/test_code_edit.h:
```diff
@@ -3445,6 +3445,98 @@ TEST_CASE("[SceneTree][CodeEdit] completion") {
+		code_edit->confirm_code_completion();
+		CHECK(code_edit->get_line(0) == "test");
+
+		/* favorise suggestion starting from the string to complete (matching start)*/
```
Make sure to keep a space before the closing `*/` (we typically wouldn't use `/* */` comment style for single line comments like this but there's a precedent in this file so it's fine to keep it consistent).

ajreckof committed:
sort code completions with rules

Fixups

Add levenshtein distance for comparisons, remove kind sort order, try to improve as many different use cases as possible

Trying again to improve code completion

Sort code autocompletion options by similarity based on input

To make it really brief, uses a combination `String.similiary`, the category system introduced in a previous PR, and some filtering to yield more predictable results, instead of scattering every completion option at seemingly random.

It also gives much higher priority to strings that contain the base in full, closer to the beginning or are perfect matches.

Also moves CodeCompletionOptionCompare to code_edit.cpp

Co-Authored-By: Micky <66727710+Mickeon@users.noreply.github.com>
Co-Authored-By: Eric M <41730826+EricEzaM@users.noreply.github.com>

akien-mga reviewed (approved):
Looks good to me. Let's get this tested more widely, but it should be a significant improvement compared to the Godot 4.0 autocompletion.

akien-mga commented:
Thanks!
